storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -3380,16 +3381,21 @@ public void rebalance(String topoName, RebalanceOptions options)
             options.set_principal(null);
             // check if executor counts are correctly specified
             StormTopology stormTopology = tryReadTopologyFromName(topoName);
-            Set<String> comps = new HashSet<>();
+            Set<String> comps = new TreeSet<>();
             comps.addAll(stormTopology.get_spouts().keySet());
             comps.addAll(stormTopology.get_bolts().keySet());
             Map<String, Integer> execOverrides = options.is_set_num_executors() ? options.get_num_executors() : Collections.emptyMap();
             for (Map.Entry<String, Integer> e: execOverrides.entrySet()) {
                 String comp = e.getKey();
-                Integer value = e.getValue();
-                if (!comps.contains(comp)) {
-                    throw new WrappedInvalidTopologyException(String.format("Component %s does not exist in topology %s", comp, topoName));
+                // validate non-system component ids
+                if (!Utils.isSystemId(comp) && !comps.contains(comp)) {
```

Comment:
Maybe we can follow the code at https://github.com/apache/storm/blob/master/storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java#L2638-L2639 to get all the components?

Reply:
Never mind. I think it is okay to not check system component for now.